### PR TITLE
ccode : implemented ccode for diracdelta and heaviside functions

### LINKED ISSUE #14307: Not supported in C: Heaviside, DiracDelta
Hi,

When generating code using ccode, sympy does not translate the Heaviside and DiracDelta function, but instead inserts a comment in code:
```
// Not supported in C:
// Heaviside
```
FWIW, a C translation might look like this:

```
static inline double Heaviside (double x) { return x < 0 ? 0 : x > 0 ? 1 : 0.5; }
static inline double DiracDelta (double x) { return x == 0 ? +INFINITY : 0; }
```

### PR DESCRIPTION
**Pull request by jashan498**
Fixes #14307

please review @moorepants @bjodah

### CODE REVIEW

**Comment by asmeurer** [comment]
I don't think it's particularly useful to have DiracDelta. Chances are if you are generating code that has DiracDelta you have a bug somewhere.

File: sympy/printing/ccode.py [hunk 1]
```diff
@@ -320,6 +320,14 @@ def _print_Infinity(self, expr):
     def _print_NegativeInfinity(self, expr):
         return '-HUGE_VAL'
 
+    def _print_DiracDelta(self, expr):
+        return "static inline double DiracDelta (double %(s)s) \
+        { return %(s)s == 0 ? HUGE_VAL : 0; }" % {'s' : self._print(*expr.args)}
+
+    def _print_Heaviside(self, expr):
+        return "static inline double Heaviside (double %(s)s) \
```
Comment (asmeurer) [inline sympy/printing/ccode.py]: Why is this a function definition? Can this be used as an expression?

Comment (jashan498) [reply sympy/printing/ccode.py]: Oh, I thought full function definition is to be printed. 
So should I remove `_print_DiracDelta` and change the other print function to return just an expression (or Heaviside is unnecessary too)?

**Comment by Siesh1oo** [comment]
@asmeurer: those are generated from sympy itself when symbolically differentiating a function containing sympy.Min()/Max() and emitting ccode() from the result.

**Comment by Siesh1oo** [comment]
@asmeurer @jashan498: I am wondering whether these functions wouldn’t better get replaced by expressions in the sympy expression tree, before emitting code, since at least DiracDelta would copy the expression tree 3 times if written as naive C/Fortran expression (and they do not map to a builtin function of the common math libraries anyways).

Then optimizers like sympy.cse() and simplify could take advantage of the available information and would need to evaluate the subexpression only once.

**Comment by Siesh1oo** [comment]
Another note on the use of HUGE_VAL in the ccode() generator.

The +/-HUGE_VAL bit pattern is not semantically equivalent to +/-INFINITY in IEEE 754. It is reserved as error indicator for ERANGE in POSIX C and C++. True +/-inf with proper overflow and propagation is defined as +/-INFINITY in <math.h>

——
> HUGE_VAL
> Huge value
> Macro constant that expands to a positive expression of type double.
> 
> A function returns this value when the result of a mathematical operation yields a value that is too large in magnitude to be representable with its return type. This is one of the possible range errors, and is signaled by setting errno to ERANGE.
> 
> Actually, functions may either return a positive or a negative HUGE_VAL (HUGE_VAL or -HUGE_VAL) to indicate the sign of the result.

**Comment by bjodah** [comment]
Yes, we should use `_print_Infinity`. (which uses `INFINITY` in the C99CodePrinter). I think it isn't guaranteed to be defined for earlier standards? (http://en.cppreference.com/w/c/numeric/math/INFINITY)

**Comment by Siesh1oo** [comment]
The #define INFINITY had been introduced as standard with C99, the IEEE754-1985 bit representation used by processors and math libraries had been around since the 1980s... so it might be required to #define INFINITY locally to the right bit pattern if compiling against pre-C99 system headers.

**Comment by Siesh1oo** [comment]
My memory was fooling me about HAVE_VAL/infinity in POSIX, they actually evaluate the same in C99: http://pubs.opengroup.org/onlinepubs/007904975/basedefs/math.h.html. 

Please ignore the remark above wrt this; I am sorry about this distraction from the primary issue in this thread.

**Comment by moorepants** [comment]
@jashan498 This needs some updates or responses to the comments.

**Pull request closed without merging**